cpp-terminal/platforms/exception.hpp

```diff
@@ -0,0 +1,30 @@
+/*
+* cpp-terminal
+* C++ library for writing multiplatform terminal applications.
+*
+* SPDX-FileCopyrightText: 2019-2023 cpp-terminal
+*
+* SPDX-License-Identifier: MIT
+*/
+
+#include "cpp-terminal/exception.hpp"
+
+namespace Term
+{
+
+namespace Private
+{
+
+// Helper for windows errors
+#if defined(_WIN32)
+class WindowsError : public Term::Exception
+{
+public:
+  WindowsError(const unsigned long& error);
+  virtual ~WindowsError() = default;
+};
+#endif
```

Comment:
Good practice in C++ is to keep all platform specific things in one "cpp" file. And the rest of the code to be fully platform independent, no ifdefs, and no ifdefs/macros in header files.

That's how cpp-terminal was originally written as well.

Is there a way to design this PR in a similar way?

Reply:
@certik I fully agree with this and I tried to do it as much as possible. I try to find a way to avoid such things but this class is really windows specific. I could make it an useless class in Unix, but then it could be used by mistake in Unix specific code and could be har to debug. Please note that this is purely internal class and is only use by the developpers of cpp-terminal not the user (the .hpp is not installed) and it is not part of the public interface of the library. 

From a user perspective this class doesn't exist and he will catch the exception with the general Term::Exception. I designed it this way to be sure the class is not used in UNIX (code will not compile because the class would not exist) and to keep it internal to avoid user to use it in his own code.

The same apply to to_wide and to_narrow it is very specifics to Windows as the best approach up to know to have unicode is to use the *W functions that require `std::wstring` encoded in utf16. This function are purely internal too and trying to export then to UNIX would be useless and dangerous because wchar_t is 16bits in windows and 32bits in UNIX to my understanding. Puting then inside `#if defined(_WIN32)` assures this class is not available to UNIX world and as such gives a hint to the developper on the internal core library that he is certainly doing something wrong if he needs this function in UNIX.

If you really want to avoif the `#ifdef` in .h then the only possibility is to make them a dumb function in UNIX maybe with some `static_assert` or something similar.